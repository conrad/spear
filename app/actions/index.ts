export * from './searchForm';
export * from './searches';
export * from "./results";
