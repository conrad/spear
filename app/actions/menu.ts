import { actionCreatorVoid } from './helpers';

export const toggleMenu = actionCreatorVoid('TOGGLE_MENU');

export function toggleMenuDropDown(): Function {
  return (dispatch: Function) => {
    dispatch(toggleMenu());
  }
}
