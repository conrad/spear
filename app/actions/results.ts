import { actionCreator, actionCreatorVoid } from './helpers';
import TextDocWriter from "../local/export/resultsExporter"; 
import * as constants from '../constants'
import IResults from '../types/IResults';

export interface IExcerptIndices {
  resultIndex: number,
  excerptIndex: number,
}

export const saveResults = actionCreator<IResults>('SAVE_RESULTS')
export const toggleShowResultsWindow = actionCreatorVoid(constants.TOGGLE_SHOW_RESULTS_WINDOW)
export const toggleShowResult = actionCreator<number>(constants.TOGGLE_SHOW_RESULT)
export const showOverlay = actionCreator<IExcerptIndices>(constants.SHOW_RESULT_OVERLAY)
export const hideOverlay = actionCreatorVoid(constants.HIDE_RESULT_OVERLAY)
export const exportResults = actionCreatorVoid(constants.EXPORT_RESULTS)

export function setResults(results: IResults) {
  return (dispatch: Function) => {
    dispatch(saveResults(results))
  }
}

export function toggleShowSearchResultRows(index: number) {
  return (dispatch: Function) => {
    dispatch(toggleShowResult(index))
  }
}

export function showResultOverlay(resultIndex: number, excerptIndex: number) {
  return (dispatch: Function) => {
    dispatch(showOverlay({ resultIndex, excerptIndex }))
  }
}

export function hideResultOverlay() {
  return (dispatch: Function) => {
    dispatch(hideOverlay())
  }
}

export function saveResultsToFile() {
  return (dispatch: Function, getState: Function) => {
    const { searches, results } = getState()
    const textDocWriter = new TextDocWriter()
    textDocWriter.saveResults("Spear_Search_Results.txt", searches.searches, results.items)
    dispatch(exportResults())
  }
}
