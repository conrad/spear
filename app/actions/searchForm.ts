import { actionCreatorVoid, actionCreator } from './helpers';

export const submitSearch = actionCreatorVoid('SUBMIT_SEARCH');
export const unsetPhrase = actionCreator<number>('DELETE_PHRASE');
export const resetFile = actionCreatorVoid('RESET_FILE');
export const setFile = actionCreator<string>('SET_FILE');
export const appendPhrase = actionCreator<string>('APPEND_PHRASE');
export const setAddingPhrase = actionCreator<boolean>('SET_ADDING_PHRASE');

export function addFile(filename: string) {
  return (dispatch: Function, getState: Function) => {
    // const { searchForm } = getState();
    dispatch(setFile(filename));
  }
}

export function addPhrase(phrase: string) {
  return (dispatch: Function, getState: Function) => {
    // const { searchForm } = getState();
    console.log('addphrase in actions');
    dispatch(appendPhrase(phrase));
    dispatch(setAddingPhrase(false));
  }
}

export function deletePhrase(index: number) {
  return (dispatch: Function, getState: Function) => {
    // const { searchForm } = getState();
    dispatch(unsetPhrase(index));
  }
}

export function startAddingPhrase() {
  return (dispatch: Function) => {
    dispatch(setAddingPhrase(true));
  };
}

export function submitValidSearch() {
  return (dispatch: Function, getState: Function) => {
    const { searchForm } = getState();

    if (!searchForm.filename || searchForm.phrases.length < 1) {
      return;
    }

    dispatch(submitSearch());
  };
}
