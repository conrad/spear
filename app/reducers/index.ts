import { combineReducers, Reducer } from 'redux';
import { routerReducer as routing } from 'react-router-redux';
import counter, { TState as TCounterState } from './counter';
import searchForm, { TState as TSearchFormState } from './searchForm';

const rootReducer = combineReducers({
  counter,
  searchForm,
  routing: routing as Reducer<any>
});

export interface IState {
  counter: TCounterState;
  searchForm: TSearchFormState;
}

export default rootReducer;
