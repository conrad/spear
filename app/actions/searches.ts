import { actionCreator, actionCreatorVoid } from './helpers';
import { IPhrase, ISearch, IMove } from '../reducers/searches';
import JsonReader from '../local/import/searchLoader';
import JsonWriter from '../local/export/searchExporter';

export const storeSearch = actionCreator<ISearch>('STORE_SEARCH');
export const deleteSearch = actionCreator<ISearch>('DELETE_SEARCH');
export const moveSearch = actionCreator<IMove>('MOVE_SEARCH');
export const unsetPhrase = actionCreator<IPhrase>('UNSET_PHRASE');
export const toggleSearchAsUsed = actionCreator<number>('TOGGLE_SEARCH_AS_USED');
export const setNewSearchName = actionCreator<string>('SET_NEW_SEARCH_NAME');
export const storeNewSearch = actionCreatorVoid('STORE_NEW_SEARCH');
export const setActiveSearch = actionCreator<number>('SET_ACTIVE_SEARCH');
export const storeSearchesFromProfile = actionCreator<Array<ISearch>>('STORE_SEARCHES_FROM_PROFILE');
export const exportSearchesToFile = actionCreatorVoid('EXPORT_SEARCHES_TO_FILE') 

export function addSearch(search: ISearch) {
  return (dispatch: Function) => {
    dispatch(storeSearch(search));
  }
}

export function updateSearch(search: ISearch) {
  return (dispatch: Function) => {
    dispatch(storeSearch(search));
  }
}

export function reorderSearch(move: IMove) {
  return (dispatch: Function) => {
    dispatch(moveSearch(move));
  };
}

export function deletePhrase(phraseIndex: number, searchIndex: number) {
  return (dispatch: Function) => {
    const phrase: IPhrase = {
      index: phraseIndex,
      text: '',
      searchIndex,
    };
    
    dispatch(unsetPhrase(phrase));
  }    
}

export function setSearchAsUsed(index: number, isUsed: boolean|null) {
  return (dispatch: Function) => {
    dispatch(toggleSearchAsUsed(index));
  };
}

export function updateNewSearchName(name: string) {
  return (dispatch: Function) => {
    dispatch(setNewSearchName(name));
  };    
}

export function exportSearches() {
  return (dispatch: Function, getState: Function) => {
    const { searches } = getState();
    const jsonWriter = new JsonWriter();
    try {
      jsonWriter.saveProfile('searches_profile.json', searches);
    } catch (e) {
      console.log('Error saving searches:', e);
    }
    // dispatch(exportSearchesToFile());
  }
}

export function addSearchesFromProfile(file: File) {
  const searches: Array<ISearch> = JsonReader.retrieveSearchesFromFile(file.path);

  return (dispatch: Function) => {
    dispatch(storeSearchesFromProfile(searches));
  };    
}

export function addNewSearch() {
  return (dispatch: Function) => {
    dispatch(storeNewSearch());
  };     
}

export function selectSearch(index: number) {
  return (dispatch: Function) => {
    dispatch(setActiveSearch(index));
  };     
}

export function removeSearch(index: number, name: string) {
  return (dispatch: Function) => {
    const search = {
      index,
      name,
      phrases: [],
      isIncluded: false,
      isEditing: false
    };
    dispatch(deleteSearch(search));
  };
}
