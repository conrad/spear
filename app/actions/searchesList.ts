import { actionCreator } from './helpers';
import { ISearch, IMove } from '../reducers/searchesList';

export const storeSearch = actionCreator<ISearch>('STORE_SEARCH');
export const deleteSearch = actionCreator<ISearch>('DELETE_SEARCH');
export const moveSearch = actionCreator<IMove>('MOVE_SEARCH');

export function addSearch(index: number, name: string, description: string|null, phrases: Array<string>) {
  return (dispatch: Function) => {
    const search = {
      index,
      name,
      description,
      phrases,
      isIncluded: false,
      isEditing: false,
    };

    dispatch(storeSearch(search));
  }
}

export function reorderSearch(move: IMove) {
  return (dispatch: Function) => {
    dispatch(moveSearch(move));
  };
}

export function removeSearch(index: number, name: string) {
  return (dispatch: Function) => {
    const search = {
      index,
      name,
      description: '',
      phrases: [],
      isIncluded: false,
      isEditing: false
    };
    dispatch(deleteSearch(search));
  };
}
