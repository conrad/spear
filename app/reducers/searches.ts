import { IAction } from '../actions/helpers';
import { clone } from '../utils/helpers';
import { storeSearch, deleteSearch, unsetPhrase } from '../actions/searches';
import { submitSearch, setNewPhrase, setIsPhraseUsed, setFile, resetFile } from '../actions/searchForm';


const initialState: ISearchesState = {
  searches: [{
    index: 0,
    name: "Commerce Clause",
    phrases: ['hey ma', 'whats up'],
    isIncluded: true,
    isEditing: false,
  }],
  currentSearchIndex: 0,
  newSearch: null,
  isNewSearchUsed: false,
  filename: "",
  isValidFile: true,
  newPhrase: "",
  isNewPhraseUsed: false,
}; 

export interface ISearchesState {
  searches: Array<ISearch>,
  currentSearchIndex: number,
  newSearch: string|null,
  isNewSearchUsed: boolean,
  filename: string,
  isValidFile: boolean,
  newPhrase: string,
  isNewPhraseUsed: boolean,
};

export interface ISearch {
  index: number,
  name: string,
  description?: string,
  phrases: Array<string>,
  isIncluded: boolean,
  isEditing: boolean,
};

export interface IMove {
  initialIndex: number,
  nextIndex: number,
};

export interface IPhrase {
  index: number,
  text: string,
  searchIndex: number
};

export type TState = ISearchesState;

export default function searches(state: ISearchesState = initialState, action: IAction) {
  let newState: ISearchesState = clone(state);

  if (storeSearch.test(action)) {
    newState.searches[action.payload.index] = action.payload;
    return newState;

  } else if (unsetPhrase.test(action)) {
    newState.searches[action.payload.searchIndex].phrases.splice(action.payload.index, 1);
    return newState;

  } else if (deleteSearch.test(action)) {
    let index: number|null = null;
    newState.searches.map((search, i) => {
      if (search.name === action.payload.name)
        index = i;
    });
    if (typeof index !== 'number') {
      console.log('Search name did not match a pre-existing search to delete.');
      return newState;
    } 
    newState.searches.splice(index, 1);
    return newState;    

  } else if (setNewPhrase.test(action)) {
    newState.newPhrase = action.payload;
    return newState;

  } else if (setIsPhraseUsed.test(action)) {
    newState.isNewPhraseUsed = action.payload;
    return newState;

  } else if (setFile.test(action)) {
    newState.isValidFile = true;
    newState.filename = action.payload;
    return newState;

  } else if (resetFile.test(action)) {
    newState.isValidFile = false;
    newState.filename = '';
    return newState;

  } else if (submitSearch.test(action)) {
    console.log("filename:", newState.filename);
    // console.log("first phrase:", newState.search.phrases[0]);
    return newState;
  }

  return newState;
}


// import { IAction } from '../actions/helpers';
// import { submitSearch, setNewPhrase, setIsPhraseUsed, setFile, resetFile } from '../actions/searchForm';
// import { clone } from '../utils/helpers';


// const initialState = { 
//   filename: "",
//   isValidFile: true,
//   phrases: ['hey ma', 'whats up'],
//   newPhrase: "",
//   isNewPhraseUsed: false
// };

// export interface IFormState {
//   filename: string,
//   isValidFile: boolean,
//   phrases: Array<string>,
//   newPhrase: string,
//   isNewPhraseUsed: boolean
// };

// export interface IPhrase {
//   index: number,
//   text: string,
//   searchIndex: number
// };

// export type TState = IFormState;

// export default function searchForm(state: IFormState = initialState, action: IAction) {
//   let newState: IFormState = clone(state);

//   if (appendPhrase.test(action)) {
//     newState.phrases.push(action.payload);
//     return newState;
//   } else if (setNewPhrase.test(action)) {
//     newState.newPhrase = action.payload;
//     return newState;
//   } else if (unsetPhrase.test(action)) {
//     newState.phrases.splice(action.payload, 1);
//     return newState;
//   } else if (setIsPhraseUsed.test(action)) {
//     newState.isNewPhraseUsed = action.payload;
//     return newState;
//   } else if (setFile.test(action)) {
//     newState.isValidFile = true;
//     newState.filename = action.payload;
//     return newState;
//   } else if (resetFile.test(action)) {
//     newState.isValidFile = false;
//     newState.filename = '';
//     return newState;
//   } else if (submitSearch.test(action)) {
//     console.log("filename:", newState.filename);
//     console.log("first phrase:", newState.phrases[0]);
//     return newState;
//   }

//   return newState;
// }
