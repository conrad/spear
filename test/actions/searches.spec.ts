import { spy } from 'sinon';
import * as actions from '../../app/actions/searches';
import { ISearch, IMove } from '../../app/reducers/searches';

const mockSearch: ISearch = {
  isEditing: false,
  isIncluded: false,
  name: '',
  phrases: [],
};

const mockMove: IMove = {
  initialIndex: 1,
  nextIndex: 2
};

describe('actions', () => {
  it('should storeSearch should create storeSearch action', () => {
    expect(actions.storeSearch(mockSearch)).toMatchSnapshot();
  });

  it('should deleteSearch should create deleteSearch action', () => {
    expect(actions.deleteSearch(mockSearch)).toMatchSnapshot();
  });

  it('should moveSearch should create moveSearch action', () => {
    expect(actions.moveSearch(mockMove)).toMatchSnapshot();
  });

  it('should saveSearch should create storeSearch action', () => {
    const fn = actions.addSearch(
      mockSearch.index, 
      mockSearch.name, 
      mockSearch.description, 
      mockSearch.phrases
    );
    expect(fn).toBeInstanceOf(Function);
    const dispatch = spy();
    fn(dispatch);
    expect(dispatch.calledWith({ 
      type: actions.storeSearch.type, 
      payload: mockSearch
    })).toBe(true);
  });

  it('should removeSearch should create deleteSearch action', () => {
    const fn = actions.removeSearch(
      mockSearch.index, 
      mockSearch.name
    );
    expect(fn).toBeInstanceOf(Function);
    const dispatch = spy();
    fn(dispatch);
    expect(dispatch.calledWith({ 
      type: actions.deleteSearch.type, 
      payload: {
        index: mockSearch.index,
        name: mockSearch.name,
        description: '',
        phrases: [],
        isEditing: false
      }
    })).toBe(true);
  });

  it('should reorderSearch should create moveSearch action', () => {
    const fn = actions.reorderSearch(mockMove);
    expect(fn).toBeInstanceOf(Function);
    const dispatch = spy();
    fn(dispatch);
    expect(dispatch.calledWith({ 
      type: actions.moveSearch.type, 
      payload: mockMove
    })).toBe(true);
  });
});