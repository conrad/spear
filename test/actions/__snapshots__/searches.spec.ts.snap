// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`actions should deleteSearch should create deleteSearch action 1`] = `
Object {
  "payload": Object {
    "index": 0,
    "isEditing": false,
    "isIncluded": false,
    "name": "",
    "phrases": Array [],
  },
  "type": "DELETE_SEARCH",
}
`;

exports[`actions should moveSearch should create moveSearch action 1`] = `
Object {
  "payload": Object {
    "initialIndex": 1,
    "nextIndex": 2,
  },
  "type": "MOVE_SEARCH",
}
`;

exports[`actions should storeSearch should create storeSearch action 1`] = `
Object {
  "payload": Object {
    "index": 0,
    "isEditing": false,
    "isIncluded": false,
    "name": "",
    "phrases": Array [],
  },
  "type": "STORE_SEARCH",
}
`;

exports[`actions should unsetPhrase should create unsetPhrase action 1`] = `
Object {
  "payload": Object {
    "index": 0,
    "searchIndex": 0,
    "text": "",
  },
  "type": "UNSET_PHRASE",
}
`;

exports[`searches actions should deleteSearch should create deleteSearch action 1`] = `
Object {
  "payload": Object {
    "index": 0,
    "isEditing": false,
    "isIncluded": false,
    "name": "",
    "phrases": Array [],
  },
  "type": "DELETE_SEARCH",
}
`;

exports[`searches actions should moveSearch should create moveSearch action 1`] = `
Object {
  "payload": Object {
    "initialIndex": 1,
    "nextIndex": 2,
  },
  "type": "MOVE_SEARCH",
}
`;

exports[`searches actions should storeSearch should create storeSearch action 1`] = `
Object {
  "payload": Object {
    "index": 0,
    "isEditing": false,
    "isIncluded": false,
    "name": "",
    "phrases": Array [],
  },
  "type": "STORE_SEARCH",
}
`;

exports[`searches actions should unsetPhrase should create unsetPhrase action 1`] = `
Object {
  "payload": Object {
    "index": 0,
    "searchIndex": 0,
    "text": "",
  },
  "type": "UNSET_PHRASE",
}
`;
